### PR TITLE
azurerm_container_app - Allow multiple custom_domain blocks within the ingress block

### LINKED ISSUE #20784: azurerm_container_app - Does not allow multiple custom_domain blocks
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.3.5

### AzureRM Provider Version

3.45.0

### Affected Resource(s)/Data Source(s)

azurerm_container_app

### Terraform Configuration Files

```hcl
# Relevant section of terraform with dummy values

resource "azurerm_container_app" "container_app" {
  name                         = "container-app"
  container_app_environment_id = "<env_id>"
  resource_group_name          = "container-app-rg"
  revision_mode                = "Single"

  ingress {
    target_port = 80
    external_enabled = true
    traffic_weight {
      percentage = 100
    }
    custom_domain {
      certificate_id = "<certificate_id_from_container_apps_environment>"
      name          = "customDomain1"
    }
    custom_domain {
      certificate_id = "<certificate_id_from_container_apps_environment>"
      name          = "customDomain2"
    }
  }

  registry {
    server               = "containerappeun.azurecr.io"
    identity             = "<user_assigned_identity>"
  }

  identity {
    type = "UserAssigned"
    identity_ids = ["<user_assigned_identity>"]
  }

  template {
    container {
      name   = "container-app"
      image  = "<valid_image>"
      cpu    = var.container_cpu_request
      memory = var.container_mem_request
    }
    min_replicas = var.container_min_replicas
    max_replicas = var.container_max_replicas
  }
}
```

### Debug Output/Panic Output

```shell
│ Error: Too many custom_domain blocks
│ 
│   on main.tf line 127, in resource "azurerm_container_app" "container_app":
│  127:     custom_domain {
│ 
│ No more than 1 "custom_domain" blocks are allowed
```

### Expected Behaviour

According to the documentation:
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/container_app#custom_domain

One or more custom_domain block should be allowed and hence when applying terraform,
Container App should be created with 2 custom domains configured.

### Actual Behaviour

terraform plan fails with the following validation error:
```
│ Error: Too many custom_domain blocks
│ 
│   on main.tf line 127, in resource "azurerm_container_app" "container_app":
│  127:     custom_domain {
│ 
│ No more than 1 "custom_domain" blocks are allowed
```

Looking at the code (With my very limited knowledge of Go)
In the customDomain schema `maxitems` seems to be set to one which could be the problem.

### Steps to Reproduce

`terraform plan`

### Important Factoids

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by srjobinar**
According to the azurerm_container_app documentation multiple custom domain blocks are allowed in the ingress block.
image

Currently terraform plan validation fails when multiple blocks are configured.

The following changes are made:
1. Recreate the test certificate with CN `acceptancetest.contoso.net` and SAN `acceptancetest.contoso.net`
2. Add tests which creates container Apps with multiple custom_domain block within ingress block
3. Remove the maxItems schema validation for custom_domain schema

fixes #20784

### CODE REVIEW

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @srjobinar, this change looks good though the test you've made is failing. Once that's fixed up, we can merge this in

```
=== CONT  TestAccContainerAppResource_basicWithCustomDomains
testcase.go:110: Step 1/2 error: Error running apply: exit status 1
Error: creating Container App (Subscription: "*******"
Resource Group Name: "acctestRG-CAEnv-230308223112188763"
Container App Name: "acctest-capp-230308223112188763"): performing CreateOrUpdate: containerapps.ContainerAppsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidCustomHostNameValidation" Message="A TXT record pointing from asuid.acceptancetest.contoso.com to 633CA537530783A9D6C4F655280B4FACA5D10CD906F227C70F7B5C7DD73DE026 was not found."
with azurerm_container_app.test,
on terraform_plugin_test.tf line 48, in resource "azurerm_container_app" "test":
48: resource "azurerm_container_app" "test" {
creating Container App (Subscription: "*******"
Resource Group Name: "acctestRG-CAEnv-230308223112188763"
Container App Name: "acctest-capp-230308223112188763"): performing
CreateOrUpdate: containerapps.ContainerAppsClient#CreateOrUpdate: Failure
sending request: StatusCode=0 -- Original Error:
Code="InvalidCustomHostNameValidation" Message="A TXT record pointing from
asuid.acceptancetest.contoso.com to
633CA537530783A9D6C4F655280B4FACA5D10CD906F227C70F7B5C7DD73DE026 was not
found."
--- FAIL: TestAccContainerAppResource_basicWithCustomDomains (1918.67s)
```

**Comment by srjobinar** [comment]
Hey @mbfrahry, thanks for checking this out. 
I am currently able to create container apps using azapi resource using a similar config. The only difference being that the managed environment is within a VNet for my usecase. 

Was wondering if this error occurs only for a managedEnv which is NOT within a VNet.

Additionally I can see that there is already an issue open with Microsoft for the same problem:
https://github.com/microsoft/azure-container-apps/issues/525

**Comment by jrnk** [comment]
Is this a bugfix or is this a newly added feature since the [current documentation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/container_app#custom_domain) states: `"(Optional) One or more custom_domain block as detailed below."` 

However, upon trying to apply this configuration we get an error:
```│ Error: Too many custom_domain blocks
│ No more than 1 "custom_domain" blocks are allowed```

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @srjobinar 
Thanks for this PR.  It looks mostly good to me, but we'll need you to remove the changed binary file (testacc.pfx) from the PR please, and I'll get the tests run and a final review completed.

Thanks!

**Comment by srjobinar** [comment]
> Hi @srjobinar Thanks for this PR. It looks mostly good to me, but we'll need you to remove the changed binary file (testacc.pfx) from the PR please, and I'll get the tests run and a final review completed.
> 
> Thanks!

@jackofallops 
The updated file was to add multiple domains in the certificate, if I remove that how do I test the change to allow multiple domains?

File: internal/services/containerapps/helpers/container_apps.go
Comment (slessardjr) [inline internal/services/containerapps/helpers/container_apps.go:224]: Based on the documentation, instead of removing the line, it looks like it might have been a typo.  Should this be `MinItems : 1,` based on the documentation stating:

> One or more custom_domain block as detailed below.

**Comment by jackofallops** [comment]
> Hey @srjobinar, this change looks good though the test you've made is failing. Once that's fixed up, we can merge this in
> 
> ```
> === CONT  TestAccContainerAppResource_basicWithCustomDomains
> testcase.go:110: Step 1/2 error: Error running apply: exit status 1
> Error: creating Container App (Subscription: "*******"
> Resource Group Name: "acctestRG-CAEnv-230308223112188763"
> Container App Name: "acctest-capp-230308223112188763"): performing CreateOrUpdate: containerapps.ContainerAppsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidCustomHostNameValidation" Message="A TXT record pointing from asuid.acceptancetest.contoso.com to 633CA537530783A9D6C4F655280B4FACA5D10CD906F227C70F7B5C7DD73DE026 was not found."
> with azurerm_container_app.test,
> on terraform_plugin_test.tf line 48, in resource "azurerm_container_app" "test":
> 48: resource "azurerm_container_app" "test" {
> creating Container App (Subscription: "*******"
> Resource Group Name: "acctestRG-CAEnv-230308223112188763"
> Container App Name: "acctest-capp-230308223112188763"): performing
> CreateOrUpdate: containerapps.ContainerAppsClient#CreateOrUpdate: Failure
> sending request: StatusCode=0 -- Original Error:
> Code="InvalidCustomHostNameValidation" Message="A TXT record pointing from
> asuid.acceptancetest.contoso.com to
> 633CA537530783A9D6C4F655280B4FACA5D10CD906F227C70F7B5C7DD73DE026 was not
> found."
> --- FAIL: TestAccContainerAppResource_basicWithCustomDomains (1918.67s)
> ```

Hi @srjobinar - After re-review, this does indeed have a circular dependency as the value for `custom_domain_verification_id` isn't known until after the app is created, so I think it will need to be implemented in a different way. Specifically, I think this will need to be a new association / meta resource (something like `azurerm_container_app_custom_domain`) that updates the `azurerm_container_app` after creation, so that the `CustomDomainVerificationId` can be known and appropriate `TXT` record(s) created in the relevant DNS Zone(s), then the meta resource can be used to update the app appropriately. Is this something you feel you could do? If not, it's something we can try and schedule in for the team to take a look into?

If it helps, I think the new resource would look something like:

```terraform
resource "azurerm_container_app_custom_domain" "example" {
  name                                     = "my-custom-host.mydomain.net"
  container_app_id                         = azurerm_container_app.example.id
  container_app_environment_certificate_id = azurerm_container_app_environment_certificate.example.id
  custom_domain_verification_record_id     = azurerm_dns_txt_record.example.id
}
```

There may need to be some additional consideration for cases where the DNS record isn't in Azure DNS?

**Comment by srjobinar** [comment]
```
Is this something you feel you could do? If not, it's something we can try and schedule in for the team to take a look into?
```
Hey @jackofallops

I believe creating a new resource from scratch would be a bit hard for me considering my very limited experience in golang and other deliverables I have. Please do go ahead with the change :)

Just one point for you to consider maybe:
I was able to create multiple custom domains for container apps within a Vnet in Azure using terraform. Used azapi resource at the time for doing that. So there are two possibilities here:
  1. azapi resource somehow is not affected by this circular dependency. If so why?
  2. This circular dependency is not affecting container apps created within a Vnet, if so why?

Might be worth for you to check this as well. I'll try and find some time to redo the above step using azurerm resource, which should give me more clarity on which one of the above is true

**Comment by jackofallops** [comment]
> ```
> Is this something you feel you could do? If not, it's something we can try and schedule in for the team to take a look into?
> ```
> 
> Hey @jackofallops
> 
> I believe creating a new resource from scratch would be a bit hard for me considering my very limited experience in golang and other deliverables I have. Please do go ahead with the change :)
> 
> Just one point for you to consider maybe: I was able to create multiple custom domains for container apps within a Vnet in Azure using terraform. Used azapi resource at the time for doing that. So there are two possibilities here:
> 
> 1. azapi resource somehow is not affected by this circular dependency. If so why?
> 2. This circular dependency is not affecting container apps created within a Vnet, if so why?
> 
> Might be worth for you to check this as well. I'll try and find some time to redo the above step using azurerm resource, which should give me more clarity on which one of the above is true

Hi @srjobinar - Within a Vnet, I suspect the DNS checks are skipped due to it being in local addressing? (I'd have to test this to be sure) Without a VNet, the domain must be considered public, and therefore checking DNS ownership via the validation TXT record is necessary. This limitation is not unique to AzureRM, az-api will have the same constraint / failure without a Vnet as the `CustomDomainVerificationId` is not known at creation time so the DNS validation will fail. Adding custom domain configuration as updates _after_ the resource is created would be possible since the value would be known, and the appropriate TXT record could be added. However, unless Terraform/AzureRM is aware of these changes, due to the nature of the Azure API these can be removed even if configured using the AzApi provider. Since we must cater for an `apply` to complete all tasks in one execution, we need to be able to create the container app followed by the DNS record for `CustomDomainVerificationId`, followed in turn by adding the custom domains to the container app. The only way to do this in Terraform is to have a discrete resource for the custom domain so the original resource can be modified appropriately.

Whilst we could update the resource to allow for multiple custom domains in the existing resource, we'd only be catering for the "bring your own vnet" case, so we would need to still implement the additional association resource for users that do not. At that point, it makes more sense to have a single implementation which caters for both scenarios? WDYT?

**Comment by srjobinar** [comment]
> > This limitation is not unique to AzureRM, az-api will have the same constraint / failure without a Vnet as the CustomDomainVerificationId is not known at creation time so the DNS validation will fail

Yeah that makes sense, have confirmed the same in my tests. Just for clarity it was working with azapi in my case as the resource was created without ingress first in which the DNS record was configured via terraform. Ingress was added in subsequent runs.


> > The only way to do this in Terraform is to have a discrete resource for the custom domain so the original resource can be modified appropriately.

Just wondering since custom domain is a configuration within the ingress block, should this be a separate resource for ingress resource itself or just the custom domain?

> > At that point, it makes more sense to have a single implementation which caters for both scenarios?

Yeah I agree, would not be ideal to have to do things differently for with/without the Vnet.

**Comment by WolfspiritM** [comment]
How is the status of this? Why exactly is this waiting response?
We currently have to manually assign additional custom domains via the azure portal for months now.

I'm not exactly sure what the problem here is.
The customDomainVerificationId is also available from the environment but terraform doesn't expose that:
https://github.com/hashicorp/terraform-provider-azurerm/issues/20829
https://github.com/hashicorp/terraform-provider-azurerm/issues/20502

So IF you expose the customDomainVerificationId then domains can be provisioned with the correct records before provisioning the app. No need (yet) for an extra resource.

**Comment by teck-kcheema** [comment]
Would love to see this merged! Much needed

**Pull request closed without merging**
